ctl.c

```diff
@@ -543,10 +544,75 @@ struct ctl ctl_ldap[] = {
 };
 
 void
-ctl_symlink(char *temp, char *real, char *z)
+ctl_dhcp_local(char *temp, char *real, char *z)
 {
+	char *args[] = { PKILL, "resolvd", NULL };
+
+	if (dhcpleased_is_running()) {
+		/*
+		 * dhcpleased depends on resolvd to modify /etc/resolv.conf.
+		 * Stop resolvd to prevent further dynamic modifications.
+		 */
+		cmdargs(PKILL, args);
+		sleep(1);
+
+		/*
+		 * Local DNS management requires /etc/resolv.conf to be a
+		 * symlink to the resolv.conf.symlink file managed by nsh.
+		 */
+		if (unlink(RESOLV_CONF) == -1 && errno != ENOENT) {
+			printf("%% unlink %s: %s\n", RESOLV_CONF,
+			    strerror(errno));
+		}
+		if (symlink(RESOLVCONF_SYM, RESOLV_CONF) == -1) {
+			printf("%% symlink %s %s: %s\n", RESOLVCONF_SYM,
+			    RESOLV_CONF, strerror(errno));
+		}
+	}
+
 	rmtemp(temp);
-	symlink(real, temp);
+	if (symlink(real, temp) == -1)
+		printf("%% symlink %s %s: %s\n", real, temp, strerror(errno));
+}
+
+void
+ctl_dhcp_client(char *temp, char *real, char *z)
+{
+	char *args_stop[] = { PKILL, "resolvd", NULL };
+	char *args_start[] = { RESOLVD, NULL };
+
+	if (!dhcpleased_is_running()) {
+		/* assume dhclient is being used */
+		rmtemp(temp);
+		symlink(real, temp);
+		return;
+	}
+
+	/*
+	 * dhcpleased requires resolvd to modify /etc/resolv.conf.
+	 * As of OpenBSD 7.0, resolvd and dhcpleased are both running
+	 * by default. The /etc/resolv.conf file will not be a symlink,
+	 * but a regular file populated with nameservers proposed by
+	 * various daemons and drivers, including dhcpleased(8),
+	 * slaacd(8), sppp(4), and umb(4).
+	 */
+
+	/* Stop resolvd in case it is already running. */
+	cmdargs(PKILL, args_stop);
+	sleep(1);
+
+	/* resolvd will re-create /etc/resolv.conf on startup. */
+	if (unlink(RESOLV_CONF) == -1 && errno != ENOENT)
```

Comment:
Note that this would probably not work on flashrd.
Should it be done differently there?

There is currently no way to make resolvd use a file other than /etc/resolv.conf,
and resolvd writes this file in a way such that a symlink would be overwritten.
So essentially resolvd currently requires /etc to be read-write.